albatross/models/gp.h

```diff
@@ -59,295 +65,252 @@ template <typename FeatureType> struct GaussianProcessFit {
-    }
-
-    return MarginalDistribution(pred, marginal_variance.asDiagonal());
+  template <typename FeatureType, typename FitFeatureType, typename PredictType,
+            typename std::enable_if<has_valid_predict<ImplType, FeatureType,
+                                                      GPFitType<FitFeatureType>,
+                                                      PredictType>::value,
+                                    int>::type = 0>
+  PredictType predict(const std::vector<FeatureType> &features,
+                      const GPFitType<FitFeatureType> &gp_fit,
+                      PredictTypeIdentity<PredictType> &&) const {
+    return impl().predict(features, gp_fit, PredictTypeIdentity<PredictType>());
   }
 
-  virtual Eigen::VectorXd
-  predict_mean_(const std::vector<FeatureType> &features) const override {
-    const auto cross_cov =
-        covariance_function_(features, this->model_fit_.train_features);
-    const Eigen::VectorXd pred = cross_cov * this->model_fit_.information;
-    return pred;
-  }
+  template <
+      typename FeatureType, typename FitFeatureType, typename PredictType,
+      typename std::enable_if<
+          (!has_call_operator<CovFunc, FeatureType, FeatureType>::value ||
+           !has_call_operator<CovFunc, FeatureType, FitFeatureType>::value) &&
+              !has_valid_predict<ImplType, FeatureType,
+                                 GPFitType<FitFeatureType>, PredictType>::value,
+          int>::type = 0>
+  PredictType predict(const std::vector<FeatureType> &features,
+                      const GPFitType<FitFeatureType> &gp_fit,
+                      PredictTypeIdentity<PredictType> &&) const =
+      delete; // Covariance Function isn't defined for FeatureType.
+
+protected:
+  /*
+   * CRTP Helpers
+   */
+  ImplType &impl() { return *static_cast<ImplType *>(this); }
+  const ImplType &impl() const { return *static_cast<const ImplType *>(this); }
 
-  CovarianceFunction covariance_function_;
+  CovFunc covariance_function_;
   std::string model_name_;
 };
 
-template <typename FeatureType, typename CovFunc>
-GaussianProcessRegression<FeatureType, CovFunc>
-gp_from_covariance(CovFunc covariance_function) {
-  return GaussianProcessRegression<FeatureType, CovFunc>(covariance_function);
+template <typename CovFunc>
+class GaussianProcessRegression
+    : public GaussianProcessBase<CovFunc, GaussianProcessRegression<CovFunc>> {
+public:
+  using Base = GaussianProcessBase<CovFunc, GaussianProcessRegression<CovFunc>>;
+
+  GaussianProcessRegression() : Base(){};
+  GaussianProcessRegression(CovFunc &covariance_function)
+      : Base(covariance_function){};
+  GaussianProcessRegression(CovFunc &covariance_function,
+                            const std::string &model_name)
+      : Base(covariance_function, model_name){};
+
+  // The only reason these are here is to hide the base class implementations.
+  void fit() const = delete;
```

Comment:
Are these are intentionally hiding the `fit()` and `predict()` functions in `GaussianProcessBase`? If so, why? I would have thought we wanted those to be usable/visible.

Reply:
Yeah, it has to intentionally hide them so that `has_valid_fit` and `has_valid_predict` in base class don't end up finding themselves and causing infinite recursion!  There might be a different way around it, but this was all I could immediately think of.